File: 0053-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        // Time complexity: O(N), Space complexity: O(1)
        int maxSum = INT_MIN;
        int size = nums.size();
        int curSum = 0;
        
        for (int i = 0; i < size; i++) {
            curSum += nums[i];
            maxSum = max(curSum, maxSum);
            if (curSum < 0) {
                curSum = 0;
            }
            // maxSum = max(curSum, maxSum);
        }
        
        return maxSum;
    }
};

File: 0053-maximum-subarray/53-maximum-subarray.py
class Solution:
    # Time complexity: O(N), Space Complexity: O(1)
    def maxSubArray(self, nums: List[int]) -> int:

        if not nums:
            raise ValueError("nums must not be a non-empty list")

        max_sum = curr_sum = nums[0]

        for x in nums[1:]:
            curr_sum = max(x, curr_sum + x)
            max_sum = max(max_sum, curr_sum)
        return max_sum